src/java/com/unifina/signalpath/blockchain/SendEthereumTransaction.java

```diff
@@ -0,0 +1,545 @@
+	}
+
+	/**
+	 * wraps Web3j Resonse with utility methods
+	 */
+
+	abstract class FunctionCallResult implements ITimestamped {
+		//		String hash;
+//		Map<String, List<Object>> events;
+//		List<Object> results;
+		Date timestamp;
+		Function fn;
+
+		public FunctionCallResult(Date timestamp, Function _fn) {
+			this.timestamp = timestamp;
+			fn = _fn;
+		}
+
+		@Override
+		public Date getTimestampAsDate() {
+			return timestamp;
+		}
+
+		public abstract Response getResponse();
+
+		public Response.Error getError() {
+			return getResponse().getError();
+		}
+	}
+
+	class ConstantFunctionResult extends FunctionCallResult {
+		private EthCall ethcall;
+
+		public ConstantFunctionResult(Date timestamp, Function _fn, EthCall ec) {
+			super(timestamp, _fn);
+			ethcall = ec;
+		}
+
+		@Override
+		public Response getResponse() {
+			return ethcall;
+		}
+
+		public List<Type> getResults() {
+			return FunctionReturnDecoder.decode(
+				ethcall.getValue(), fn.getOutputParameters());
+		}
+	}
+
+	/**
+	 * non constant function call result
+	 */
+	class TransactionResult extends FunctionCallResult {
+		private EthSendTransaction web3jTx;
+
+		public TransactionResult(Date timestamp, Function _fn, EthSendTransaction tx) {
+			super(timestamp, _fn);
+			web3jTx = tx;
+		}
+
+		@Override
+		public Response getResponse() {
+			return web3jTx;
+		}
+
+		public String getTransactionHash() {
+			return web3jTx.getTransactionHash();
+		}
+
+		public Map<String, List<Object>> getEvents() throws IOException, ClassNotFoundException {
+			Map<String, List<Object>> events = new HashMap<String, List<Object>>();
+			for (EthereumABI.Event e : abi.getEvents()) {
+				ArrayList<TypeReference<?>> params = new ArrayList<TypeReference<?>>();
+				for (EthereumABI.Slot s : e.inputs) {
+					params.add(TypeReference.create(Web3jHelper.getTypeClass(s.type)));
+				}
+				Event web3jevent = new Event(e.name, params);
+				TransactionReceipt txr = Web3jHelper.getTransactionReceipt(web3j, web3jTx.getTransactionHash());
+				if (txr == null) {
+					throw new IOException("couldnt find transaction receipt for txhash: " + web3jTx.getTransactionHash());
+				}
+				List<EventValues> valueList = Web3jHelper.extractEventParameters(web3jevent, txr);
+				if (valueList == null) {
+					continue;
+				}
+				for (EventValues ev : valueList) {
+					List<Object> objs = new ArrayList<Object>();
+					List<Type> niv = ev.getNonIndexedValues();
+					for (Type t : niv) {
+						objs.add(t.getValue());
+					}
+					events.put(e.name, objs);
+				}
+			}
+			return events;
+		}
+
+	}
+
+	/**
+	 * Asynchronously handle server response, call comes from event queue
+	 *
+	 * @param event containing HttpTransaction created within sendOutput
+	 */
+	@Override
+	public void receive(FeedEvent event) {
+		if (event.content instanceof FunctionCallResult) {
+			try {
+				sendOutput((FunctionCallResult) event.content);
+			} catch (Exception e) {
+				log.error(e.getMessage());
```

Comment:
This log statement is useless. The master exception handler that catches everything will log exceptions. This exception is now logged twice.